Doc/extending/newtypes.rst

```diff
@@ -1385,143 +507,104 @@ Here is a desultory example of the implementation of the call function. ::
+performance-critical objects (such as numbers).
 
-For an object to be weakly referencable, the extension must include a
-:c:type:`PyObject\*` field in the instance structure for the use of the weak
-reference mechanism; it must be initialized to *NULL* by the object's
-constructor.  It must also set the :c:member:`~PyTypeObject.tp_weaklistoffset` field of the
-corresponding type object to the offset of the field. For example, the instance
-type is defined with the following structure::
+.. seealso::
+   Documentation for the :mod:`weakref` module.
 
-   typedef struct {
-       PyObject_HEAD
-       PyClassObject *in_class;       /* The class object */
-       PyObject      *in_dict;        /* A dictionary */
-       PyObject      *in_weakreflist; /* List of weak references */
-   } PyInstanceObject;
-
-The statically-declared type object for instances is defined this way::
-
-   PyTypeObject PyInstance_Type = {
-       PyVarObject_HEAD_INIT(&PyType_Type, 0)
-       0,
-       "module.instance",
-
-       /* Lots of stuff omitted for brevity... */
-
-       Py_TPFLAGS_DEFAULT,                         /* tp_flags */
-       0,                                          /* tp_doc */
-       0,                                          /* tp_traverse */
-       0,                                          /* tp_clear */
-       0,                                          /* tp_richcompare */
-       offsetof(PyInstanceObject, in_weakreflist), /* tp_weaklistoffset */
-   };
+For an object to be weakly referencable, the extension type must do two things:
 
-The type constructor is responsible for initializing the weak reference list to
-*NULL*::
+#. Include a :c:type:`PyObject\*` field in the C object structure dedicated to
+   the weak reference mechanism.  The object's constructor should leave it
+   *NULL* (which is automatic when using the default
+   :c:member:`~PyTypeObject.tp_alloc`).
 
-   static PyObject *
-   instance_new() {
-       /* Other initialization stuff omitted for brevity */
+#. Set the :c:member:`~PyTypeObject.tp_weaklistoffset` type member
+   to the offset of the aforementioned field in the C object structure,
+   so that the interpreter knows how to access and modify that field.
 
-       self->in_weakreflist = NULL;
+Concretely, here is how a trivial object structure would be augmented
+with the required field::
 
-       return (PyObject *) self;
-   }
+   typedef struct {
+       PyObject_HEAD
+       PyObject *weakreflist;  /* List of weak references */
+   } NoddyObject;
 
-The only further addition is that the destructor needs to call the weak
-reference manager to clear any weak references.  This is only required if the
-weak reference list is non-*NULL*::
+And the corresponding member in the statically-declared type object::
 
-   static void
-   instance_dealloc(PyInstanceObject *inst)
-   {
-       /* Allocate temporaries if needed, but do not begin
```

Comment:
This comment is important. `PyObject_ClearWeakRefs()` should be called before calling any destructors.

Reply:
Why is that?

Reply:
I suppose that this is because destructors can resurrect the deallocated object (increment the refcount) if there are weak references to it. Following call of `PyObject_ClearWeakRefs()` will fail with raising SystemError because the refcount is not 0.

Interestingly, but not all builtin types follow this rule. Seems there are bugs.